### ISSUE TITLE
Cancellation or Failure Payment will lead into fatal exception (SW 6.6.x.x)

### ISSUE THREAD

**Issue by rommelfreddy** [mollie/Shopware6#791]
If I a cancel the payment or the payment will failure, then a fatal exception will happen within the Shopware Store.

```json
{
  "errors": [
    {
      "code": "0",
      "status": "500",
      "title": "Internal Server Error",
      "detail": "Attempted to load class \u0022AsyncPaymentFinalizeException\u0022 from namespace \u0022Shopware\\Core\\Checkout\\Payment\\Exception\u0022.\nDid you forget a \u0022use\u0022 statement for another namespace?"
    }
  ]
}
```

Seems like that the Polyfill for PaymentException Class is loaded before Shopware, so the class will be Polyfill-Class will be registered before the Shopware-Class is registered. 

Personal opinion of DEV: 
I do not think it is a good idea, to have too many polyfill for Shopware. It would be better to restrict the compatibility to the Shopware versions. 
I also develop a few payment provider extensions, and we decided to do not keep the compatibility to lower SW-Versions if the codebase is too different (e.g. missing exception-classes)

Much better solution:

```php
protected function createPaymentException(string $orderTransactionId, string $errorMessage, ?\Throwable $e = null): \Throwable  
{
  if (class_exists(PaymentException::class)) {
      return PaymentException::asyncProcessInterrupted($orderTransactionId, $errorMessage, $e);
  } elseif (class_exists(SyncPaymentProcessException::class)) {
      // required for shopware version <= 6.5.3
      return SyncPaymentProcessException($orderTransactionId, $errorMessage, $e);  // @phpstan-ignore-line
  }

  // should never occur, just to be safe.
  return \RuntimeException('payment process was interrupted ' . $orderTransactionId);
}
```

**Comment by BlackScorp** [comment]
Hi thx for the info, iam going to add some checks so the polyfill class is not loaded directly.

about shopware version restriction, sadly there are too many merchants which still runs shopware 6.4.x thats why we cant drop the support. i prefer to use the polyfill classes because if we drop a version at some point, i just have to delete the polyfill classes and the code stays the same.

**Comment by rommelfreddy** [comment]
Hello @BlackScorp can you give me feedback on this please?

**Comment by BlackScorp** [comment]
@rommelfreddy sorry didnt saw the notification in github.

let me check, i cannot create different release-lines because customer want to have new payment methods and features also for older shopware version. therefore i have to make sure that the features are transfered over to different releases. 


So the class PaymentException  was added in  6.5.4.0 with NEXT-28610. Before that we had to throw each exception individually.

So lets say we have currenctly shopware 6.4 and below 6.5.4.0

the polyfill see that PaymentException does not exists but AsyncPaymentFinalizeException and CustomerCanceledAsyncPaymentException do exists. in this case, the polyfill class is loaded used during payment.

if we have shopware > 6.5.4.0 then it sees that PaymentException exists but AsyncPayment not exists so default PaymentException from shopware is used.

I have to verify with composer optimized autoloader, maybe i can reproduce the problem that our class is loaded before shopware one

**Comment by BlackScorp** [comment]
@rommelfreddy can we make a zoom/google call to check it out? maybe in your case you have a plugin installed which also have some polyfill classes or you installed mollie twice (once uploaded over admin and once via composer) 

we had once a problem where Klarna Plugin provided some polyfill classes and also we had once a problem where a customer installed multiple mollie versions.

i cannot repoduce the error you described. Even when i add following line to autoload_classmap.php in composer

`    'Shopware\\Core\\Checkout\\Payment\\PaymentException' => $vendorDir . '/custom/plugin/MolliePayments/polyfill/Shopware/core/Checkout/Payment/PaymentException.php',`

cancelling the payment still work

**Comment by rommelfreddy** [comment]
why you are using a polyfil, which got not correctly registered in the autoloader, if you could use a if/else or a simple helper to create the exception, as described within the initial Issue description. 
If you just follow the guide in the description, you will solve the problem. 

We didn't install the module twice and do not have another modules which have ugly polyfills. (i also didn't saw something like that in the past in any module - i am wondering that the shopware extension store approved a release like this)

**Comment by BlackScorp** [comment]
@rommelfreddy iam using a polyfill class so i can remove if when we drop SW 6.4 Support without any change to the source code. 

there is no need to add extra if else statement. in our pipelines we test the case in different shopware versions and the polyfills works without problems. polyfill are common way in order to provide backward compatibility, symfony for example also provides polyfill for php functions

we need to check it togethere, there should be no way that PaymentException exists but the others AsyncPaymentExceptions not. either one of them can only exists, depending on shopware version.

**Comment by BlackScorp** [comment]
@rommelfreddy class_exists and interface_exists are using autoloader (2nd parameter $autoload = true), before my class is loaded, the code checks if the other classes exists. there is no way that the class can be loaded incorrectly. i tried to install the plugin over composer on fresh shopware with composer require store.shopware.com/molliepayments, i deleted multiple times the entire vendor and run composer install or composer update. in all cases the shopware class was used if found otherwise my polyfill.

since i cannot reproduce the error, iam going to close this ticket

**Comment by rommelfreddy** [comment]
as already send by mail:

We just discovered another issue with your polyfills. The Flow-Actions does not work anymore, if your module is active, because you are creating a polyfill which registers own ActionAwares with a missing PHP annotations on the awares. 
I like the thought with the polyfils, but you can not control in which order the polyfils and Shopware got loaded.

So the only way is to remove the polyfils and solve the problem (why you have added them) in another way. (PaymentExceptions: as i have described)

Currently we have 4 different environments: DEV 1, DEV 2, Stage, Prod. Only on the environment of DEV 2 the Polyfil got NOT loaded, and everything works fine. On all other envs the module breaks the processes (of payment-exceptions and actions-awares)
We will create another patch for your extension to make sure that the extension does work. (problem with the actions-awares for the flow-builder)

------

there is no special configuration within the Shopware instance. (no feature-flags, no extensions, or something else which could change the behaviour)
The extension is added as composer dependency without uploading it via the administration (but this should not matter).

It is just a simple composer loading problem.
It is not required to have the plugin installed or else. it is only required to have your extension as dependency, and the flowbuilder does not work anymore and the cancellation of a payment (during the checkout) will throw an error. 

As already mentioned, you can not handle the loading order of classes of composer. And this will cause into the problem, that your extension MAY loaded before other classes (in this case before the payment-exception-classes and the ActionAware-classes)

Long story short: you should not use polyfils - as I described in the issue-description.

**Comment by momocode-de** [comment]
Hi! @BlackScorp!

Let me summarize the problem with the Aware interfaces again:

Let's use the [MailAware Interface](https://github.com/mollie/Shopware6/blob/master/polyfill/Shopware/Core/Framework/Event/MailAware.php) as example. I assume that you have copied the interface from Shopware 6.5 or a previous version as a polyfill into your plugin. The problem is that in Shopware 6.6 a new attribute “IsFlowEventAware” has been added to the interface, as you can see [here](https://github.com/shopware/shopware/blob/v6.6.0.0/src/Core/Framework/Event/MailAware.php). Since the attribute is missing in your polyfill, there is an error in the Flow Builder in 6.6 shops.

Now, of course, the question is why your polyfill is used in a 6.6 shop and not the original class. This is now due to the order in which the PHP files are loaded. I can't tell you exactly how composer's autoloading works, but it seems to be more or less random in which order the files are loaded. If your polyfill is now loaded before the original class, the “interface_exists” check is made in the polyfill, but since the original class is not yet loaded, the check is false and therefore your polyfill is registered as the interface.

As I said, it seems to be random, as the error occurred on our live and staging system, but not on my local system. That's probably why you can't reproduce it. But the fact is that this problem exists and you should do something about it.

**Comment by rommelfreddy** [comment]
@BlackScorp is there any updates on this breaking issue?

**Comment by BlackScorp** [comment]
@rommelfreddy will be fixed in next release

**Comment by rommelfreddy** [comment]
Hallo @BlackScorp 
sorry but this is still not a good solution. you are still using the polyfils which might bei not compatible with the current or other sw-version.

why it wont be a solution to implement a if/else statement with or without a helper, as i suggested in the issue description?

**Comment by BlackScorp** [comment]
@rommelfreddy 

the polyfills are not used in the composer.json and are not loaded automatically. they polyfills are loaded during plugin boot action. this way polyfills are loaded after shopware during runtime and not in the autoloading process of composer. so there is no way that our class is used when shopware class exists.

i will not add if/else statements because it will be hard to remove them later when we drop the support for one shopware version.
